File: string-easy/387.字符串中的第一个唯一字符.py
#
# @lc app=leetcode.cn id=387 lang=python3
#
# [387] 字符串中的第一个唯一字符
#

# @lc code=start
class Solution:
    # 104/104 cases passed (112 ms)
    # Your runtime beats 61.83 % of python3 submissions
    # Your memory usage beats 59.44 % of python3 submissions (15 MB)
    def firstUniqChar(self, s: str) -> int:
        position = dict()
        n = len(s)
        for i, ch in enumerate(s):
            if ch in position:
                position[ch] = -1
            else:
                position[ch] = i
        for i in range(n):
            if position[s[i]] != -1:
                return position[s[i]]
        return -1


# @lc code=end

class Solution_1(object):
    def firstUniqChar(self, s: str) -> int:
        import collections
        counter = collections.Counter(s)
        for i, ch in enumerate(s):
            if counter[ch] == 1:
                return i
        return -1

File: string-easy/387.字符串中的第一个唯一字符.cpp
/*
 * @lc app=leetcode.cn id=387 lang=cpp
 *
 * [387] 字符串中的第一个唯一字符
 */
#include <string>

using namespace std;

// @lc code=start
class Solution {
   public:
    // 104/104 cases passed (16 ms)
    // Your runtime beats 99.38 % of cpp submissions
    // Your memory usage beats 94.78 % of cpp submissions (10.4 MB)
    int firstUniqChar(string s) {
        int count[26]{0};
        int n = s.size();
        for (int i = 0; i != n; i++) {
            count[s[i] - 'a']++;
        }
        for (int i = 0; i != n; i++) {
            if (count[s[i] - 'a'] == 1) {
                return i;
            }
        }
        return -1;
    }
};
// @lc code=end
